core/src/tdx.rs

```diff
@@ -0,0 +1,295 @@
+        let expected = hex::encode(&report_data_expected[..report_data_expected.len().min(16)]);
+        let actual = hex::encode(&report_data[..report_data_expected.len().min(16)]);
+        return Err(RatlsError::Policy(format!(
+            "report data mismatch (nonce binding): expected prefix {expected}, quote prefix {actual}"
+        )));
+    }
+    Ok(())
+}
+
+pub fn verify_event_log_integrity(quote: &[u8], events: &[TcgEvent]) -> Result<(), RatlsError> {
+    if events.is_empty() {
+        return Err(RatlsError::Policy(
+            "event log missing from attestation response".into(),
+        ));
+    }
+
+    let parsed = parse_quote_report(quote)?;
+    let report = TdReportRef::from_report(&parsed.report)?;
+    let mut accumulator = [0u8; 48];
+    let mut last_event = None;
+    let mut processed = 0usize;
+
+    for event in events {
+        if event.imr.unwrap_or(0) != 3 {
+            continue;
+        }
+
+        let mut digest = event.digest_bytes()?;
+        if digest.len() < accumulator.len() {
+            digest.resize(accumulator.len(), 0);
+        }
+        if digest.len() != accumulator.len() {
+            let label = event.message.as_deref().unwrap_or("unknown");
+            return Err(RatlsError::Vendor(format!(
+                "event '{label}' digest has invalid length {}",
+                digest.len()
+            )));
+        }
+
+        let mut hasher = Sha384::new();
+        hasher.update(&accumulator);
+        hasher.update(&digest);
+        let hashed = hasher.finalize();
+        accumulator.copy_from_slice(&hashed);
+        processed += 1;
+        last_event = Some((
+            event.message.clone().unwrap_or_else(|| "unknown".into()),
+            hex::encode(&digest),
+        ));
+    }
+
+    if accumulator != report.as_td10().rt_mr3 {
+        let expected = hex::encode(report.as_td10().rt_mr3);
+        let actual = hex::encode(accumulator);
+        let (last_msg, last_digest) = last_event.unwrap_or_else(|| ("none".into(), "n/a".into()));
+        return Err(RatlsError::Policy(format!(
+            "event log replay does not match RTMR3 (expected {expected}, computed {actual}, events_processed {processed}, last_event '{last_msg}' digest {last_digest})"
+        )));
+    }
+
+    Ok(())
+}
+
+pub fn verify_tls_certificate_in_log(
+    events: &[TcgEvent],
+    cert_data: &[u8],
+) -> Result<(), RatlsError> {
+    let mut hasher = Sha256::new();
+    hasher.update(cert_data);
+    let cert_hash = encode(hasher.finalize());
+
+    for event in events {
+        if event
+            .message
+            .as_deref()
+            .map(|msg| msg.eq_ignore_ascii_case("New TLS Certificate"))
+            .unwrap_or(false)
+        {
+            if let Some(payload) = event.payload_as_string() {
+                if payload == cert_hash {
+                    return Ok(());
+                }
+            }
+        }
+    }
```

Comment:
Something important here is that the event log can have multiple of these events. The scenario is that the machine has been running for a long time, where it had renewed the certificate. This means that you have the first event for the old cert, and the second event for the new cert. You should therefore consider the last of these events since it's the one reflecting the currently used cert.